### ISSUE TITLE
OpenRC seedrng error

### ISSUE THREAD

**Issue by DrakoDom** [smuellerDD/lrng#24]
My system is currently running openrc-0.45.2-r2, with kernel-6.1.1.

```
seedrng                   | * Seeding random number generator ...
seedrng                   | * Seeding 512 bits and crediting
seedrng                   | * seedrng: Unable to seed: No such file or directory
seedrng                   | * Saving 512 bits of creditable seed for next boot
 [ ok ]
seedrng                   | * Error seeding random number generator
 [ !! ]
```

Is this normal behavior for OpenRC?

**Comment by smuellerDD** [comment]
Am Dienstag, 3. Januar 2023, 20:17:49 CET schrieb Dominic Esposito:

Hi Dominic,

> My system is currently running openrc-0.45.2-r2, with kernel-6.1.1.

First of all, I have not seen the following log structure so far. Thus, I am 
not sure what mechanism you use here.
> 
> seedrng                   | * Seeding random number generator ...
> seedrng                   | * Seeding 512 bits and crediting
> seedrng                   | * seedrng: Unable to seed: No such file or
> directory seedrng

This error seems to be the issue - which file is accessed? Maybe strace helps?

> | * Saving 512 bits of creditable seed
> for next boot [ ok ]
> seedrng                   | * Error seeding random number generator
>  [ !! ]
> 
> Is this normal behavior for OpenRC?

As I am not familiar with OpenRC, I cannot truly tell. But as the LRNG shall 
be a drop-in replacement for the existing /dev/random, you can easily check 
with an unpatched kernel. I.e. if the unpatched kernel does not show this 
error, it is definitely a regression in the LRNG that needs fixing.

Thanks.


Ciao
Stephan

**Comment by DrakoDom** [comment]
Hi Stephan.

My current kernel config: http://dpaste.com/88XYGKPEM
Seedrng script from OpenRC: https://dpaste.com/5P5GGN57H
Patch used for LRNG: https://github.com/CachyOS/kernel-patches/blob/master/6.1/0013-lrng.patch

I'm afraid there's no way to get OpenRC with strace.
https://github.com/ish-app/ish/issues/740#issuecomment-639150678

I am sure that seedrng is working properly on the same kernel, without the LRNG patch.

If you would like, I can provide other logs. But for now I'm not sure what would be most helpful.

**Comment by smuellerDD** [comment]
Am Mittwoch, 4. Januar 2023, 20:07:33 CET schrieb Dominic Esposito:

Hi Dominic,

> Hi Stephan.
> 
> My current kernel config: http://dpaste.com/88XYGKPEM
> Seedrng script from OpenRC: https://dpaste.com/5P5GGN57H

I just get to them now - yet I do not find these files at the server.

You also may send them to me directly to ***@***.***

> Patch used for LRNG:
> https://github.com/CachyOS/kernel-patches/blob/master/6.1/0013-lrng.patch
> 
> I'm afraid there's no way to get OpenRC with strace.
> https://github.com/ish-app/ish/issues/740#issuecomment-639150678
> 
> I am sure that seedrng is working properly on the same kernel, without the
> LRNG patch.
> 
> If you would like, I can provide other logs. But for now I'm not sure what
> would be most helpful.


Ciao
Stephan

**Comment by DrakoDom** [comment]
Hello Stephan,

Instead of using dpaste, I am using pastebin now. The files should last for 1 month.

My current kernel config: https://pastebin.com/iwvu2CkF
Seedrng script from OpenRC: https://pastebin.com/MA30TBQ9

Hopefully the files will be helpful to you now.

**Comment by smuellerDD** [comment]
Am Samstag, 7. Januar 2023, 05:22:23 CET schrieb Dominic Esposito:

Hi Dominic,

Thanks for the files. From the seedrng script, I see that the script calls a 
binary. I pulled the entire OpenRC repository and found the seedrng binary.

The error log you reported shows:

seedrng: Unable to seed: No such file or directory

To debug this, I created the binary on my system and see the same error which 
allowed me to strace it. And I found the issue. Applying patch 
ea079cc9f21e46a056d8568a10844e5ed6cbb178 fixes the issue.

Yet, as far as I see, the error should not cause a real issue on your system 
as the error is at the end of the processing and the LRNG processed the data 
as expected. Only the LRNG returned a positive integer about the number of 
bytes that were processed compared to 0 for success in the legacy /dev/random 
implementation. The patch updates the LRNG to also return 0 on success.

Thanks for the report.

Ciao
Stephan
